search-sentiments-map/test/trends-test.js

```diff
@@ -0,0 +1,243 @@
+        globalTrends: 'globalTrends1.5',
+      };
+      mockResult3 = {
+        timestamp: currentTime - TIME_RANGE_3_MS,
+        globalTrends: 'globalTrends3',
+      }
+
+      // Stub calls to the datastore.
+      sinon.stub(Datastore.prototype, 'runQuery').callsFake(() => {
+        return [mockTrendsEntries];
+      });
+    });
+
+    afterEach(() => {
+      sinon.restore();
+    });
+
+    it('should retrieve the most recent datastore entry beyond RETRIEVE_RESULTS_TIME_MS',
+        async function() {
+      let results = await trends.retrieveGlobalTrendsForTimeRange(0);
+      assert.deepEqual(results, mockResult1half);
+    });
+
+    it('should retrieve the most recent datastore entry given beyond the given time range', 
+        async function() {
+      let results = await trends.retrieveGlobalTrendsForTimeRange(2);
+      assert.deepEqual(results, mockResult3);
+    });
+  });
+
+  describe('UpdateDailyTrends', function() {
+    let mockDailyTrendsData;
+    let mockTrendsByCountry;
+    let datastoreEntities;
+    const countryJson = require('./../public/countries-with-trends.json');
+
+    beforeEach(() => {
+      datastoreEntities = [];
+
+      let mockTrendingSearches = [
+        {
+          'title': {
+            'query': 'topic1',
+            'exploreLink': 'exploreLink1',
+          },
+          'formattedTraffic': '200K+',
+          'articles': [
+            {
+              'title': 'article1',
+              'url': 'url1',
+            }, {
+              'title': 'article2',
+              'url': 'url2',
+            }
+          ],
+        }
+      ];
+      mockDailyTrendsData = JSON.stringify({
+        'default': {
+          'trendingSearchesDays': [
+            {
+              'trendingSearches': mockTrendingSearches,
+            }
+          ],
+        },
+      });
+
+      mockTrendsByCountry = [];
+      for (let i = 0; i < countryJson.length; i++) {
+        mockTrendsByCountry.push({
+          country: countryJson[i].id,
+          trends: [
+            {
+              topic: 'topic1',
+              traffic: '200K+',
+              exploreLink: 'exploreLink1',
+              articles: [
+                {
+                  title: 'article1',
+                  url: 'url1',
+                }, {
+                  title: 'article2',
+                  url: 'url2',
+                }
+              ],
+            }
+          ],
+        });
+      }
+
+      // Stub the API call to googleTrends.
+      sinon.stub(googleTrends, 'dailyTrends').resolves(mockDailyTrendsData);
+
+      // Stub the `saveTrendsAndDeletePrevious` function.
+      sinon.stub(trends, 'saveTrendsAndDeletePrevious')
+          .callsFake((trendsByCountry) => {
+        datastoreEntities.push({
+          key: 'fakeKey',
+          data: {
+            timestamp: Date.now(),
+            trendsByCountry: trendsByCountry,
+            globalTrends: trends.getGlobalTrends(trendsByCountry),
+          },
+        });
```

Comment:
Why are you stubbing this function? Unless you plan on writing a separate test for this function somewhere else it should not be stubbed. Instead, the actual datastore save function should be stubbed. This will ensure that this function works, and also the deleteAncientTrends function(which is called in the saveTrendsAndDeletePrevious function) also works. See search-test.js for examples of how to stub the datastore call itself.

Reply:
I stubbed the actual datastore save function at first, but it turned out that `await trends.updateDailyTrends();` on line 158 doesn't await for the `saveTrendsAndDeletePrevious` function and when the asserts are executed the datastoreEntities is empty. That's why I stub the function as a whole.

Reply:
So instead of stubbing `saveTrendsAndDeletePrevious`, I did the following:

```
      sinon.stub(trends, 'deleteAncientTrend').resolves('do not care');

      // Stub calls to the datastore.
      sinon.stub(Datastore.prototype, 'key').callsFake(() => {
        return 'fakeKey';
      });

      // Stub calls to the datastore.
      sinon.stub(Datastore.prototype, 'save').callsFake((entity) => {
        datastoreEntities.push(entity);
      });
```
But then `datastoreEntities` is empty

Reply:
Add an await to line 109, should fix this problem.

Reply:
fixed it by adding an "await" to the function call